### ISSUE TITLE
Term 3: Semantic Segmentation project missing information

### ISSUE THREAD

**Issue by subodh-malgonde** [udacity/sdc-issue-reports#1259]
I am doing this project right now, and I realized that there is a lot of useful information which has not been mentioned in the classroom.

1. The link for the frozen VGG16 model has been hardcoded in one of the project files `helper.py`. But this has not been mentioned any where in the classroom. The link is https://s3-us-west-1.amazonaws.com/udacity-selfdrivingcar/vgg.zip
2. The model is not a vanilla VGG16, but a fully convolutionalized version of it. Please see [this forum post](https://discussions.udacity.com/t/here-is-some-advice-and-clarifications-about-the-semantic-segmentation-project/403100/8?u=subodh.malgonde). We need to mention this in the classroom so that students don't waste a lot of time on this.
3. Points mentioned in [this post](https://discussions.udacity.com/t/here-is-some-advice-and-clarifications-about-the-semantic-segmentation-project/403100?u=subodh.malgonde).

**Comment by pierluigiferrari** [comment]
I'll summarize the points from my post here. In my opinion, all of the following points are significant omissions or misinformation in the classroom material:

1. The original FCN-8s was trained in stages, not all at once. The authors later uploaded a version that was trained all at once to their GitHub repo, but that version has one important difference: The outputs of pooling layers 3 and 4 are scaled before they are fed into the 1x1 convolutions. This fact is completely omitted in the classroom and in the project walk-through. I trained my model with and without the scaling layers and found out that the model learns much better with the scaling layers included. It didn’t converge that much faster, but I was able to reach much higher IoU and accuracy with the scaling layers. To include the scaling layers, simply add them to your model like so:

pool3_out_scaled = tf.multiply(pool3_out, 0.0001, name=‘pool3_out_scaled’)
pool4_out_scaled = tf.multiply(pool4_out, 0.01, name=‘pool4_out_scaled’)

where pool3_out and pool4_out are the outputs of the VGG-16. You then feed the scaled outputs into your 1x1 convolutions and everything is as before from there. Note that the scaling factors are not the same for the two scaling layers. I took the scaling factors directly from the Prototxt of the original Caffe implementation.

2. When adding l2-regularization, many students don't know that merely setting a regularizer in the arguments of the `tf.layers` is not enough. You also have to manually add all those regularization loss terms to your loss function, otherwise they are not doing anything. I didn’t know this at first and thought I’m doing l2-regularization when I actually wasn’t. When you look at the forums, you see that a lot of students think they are regularizing when they actually aren't. The classroom should point this out.

A few statements in the walk-through video are either confusing or incorrect:

1. Contrary to what Aaron mentions at 3:43 in the video, the pretrained VGG-16 is **not** frozen. If you load the provided VGG and build layers on top of it and train the whole thing, all weights are trainable, not only the weights of the layers you added on top.

2. Aaron mentions that the model’s output tensor needs to be reshaped into 2D because that is what TensorFlow’s softmax function requires. That is incorrect. TensorFlow’s softmax function (and tf.nn.softmax_cross_entropy_with_logits, for that matter) accepts tensors of any shape and will apply the softmax function on the last axis of the tensor. There is hence no need to reshape the output of the model, especially because you only end up reshaping it back into the image dimensions after applying softmax anyway, so it’s just clock cycles wasted for your GPU. You should NOT reshape your model output. I don’t do it in my model and it works great :slight_smile:

3. Aaron mentions that we need to add 1x1 convolutions on top of the VGG-16 network in order to preserve the spatial information of the data. That is incorrect, that is not the reason why we need to add 1x1 convs on top of the VGG-16. The pretrained VGG-16 model is already fully convolutionalized, i.e. it already contains the convolutions that replace the fully connected layers. THOSE convolutions are the ones that are used to preserve spatial information that would be lost if we kept the fully connected layers. The purpose of the 1x1 convolutions that we are adding on top of the VGG is merely to reduce the number of filters from 4096 to whatever the number of classes for our model is, THAT is the reason why we need those. This may sound pedantic, but I think it’s important to understand what is what.

**Comment by baumanab** [comment]
This is really helpful, thanks to everyone who took the time to put this information together.  The information has been added to the classroom and Github README.
